File: Algo/Leetcode/045JumpGameII.cpp
// greedy. Always find the rightmost reachable boarder.
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        if(n==0) return true;
        int step = 0;
        int right = 0;
        int end = 0;
        for(int i=0; i<nums.size()-1; i++){
            right = max(right, i+nums[i]);
            if(i == end){
                end = right;
                step++;
            }
        }
        return step;
    }
};

File: Algo/Leetcode/045JumpGameII.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_len = 0     # update max len as the most far idx currently can access
        end = 0
        count = 0
        for i, n in enumerate(nums[:-1]):       # do not run the last
            max_len = max(max_len, i+n)
            if i == end:
                end = max_len
                count += 1

        return count